## 1. Importing Libraries

In [270]:
import os

import numpy as np

import pandas as pd

## 2. Reading the Data

In [5]:
PROJECT_DIR = r'C:\Users\Sam\Desktop\Elevate_Labs\Task 1'
DATA_DIR = 'data'

In [6]:
def get_data(name):
    file_name = f'{name}.csv'
    file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)
    return pd.read_csv(file_path)

In [8]:
netflix = get_data('netflix_titles')
netflix

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

In [10]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


- The dataset contains 8,807 rows and 12 columns.
- Columns `director`, `cast`, `country`, `date_added`, `rating`, `duration` have missing values each.
- The data types of some columns isn't appropriate.

## 3. Preliminary Analysis

### 3.1 Check Data Types

In [11]:
netflix.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [13]:
netflix.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [15]:
value = netflix.date_added.iloc[6]
value

'September 24, 2021'

- `date_added` needs conversion to date_time
- `duration` column has mixed values
- it has two types of values:
- Minutes → for Movies
- Seasons → for TV Shows

## 3.2 Check for Duplicates

In [16]:
netflix.duplicated().sum()

np.int64(0)

- Fortunately, it doesn't have any duplicate rows/entries.

## 4. Detailed Analysis

### `show_id`

In [17]:
netflix.show_id

0          s1
1          s2
2          s3
3          s4
4          s5
        ...  
8802    s8803
8803    s8804
8804    s8805
8805    s8806
8806    s8807
Name: show_id, Length: 8807, dtype: object

In [18]:
(
    netflix
    .show_id
    .loc[lambda ser: ~ser.str.contains(r's[0-9]+', regex=True)]
)

Series([], Name: show_id, dtype: object)

- Format of show_id
- s{row_number}, where row_number starting from 1, 's' might be standing for 'show'
- no null values
- can just act as an unique identifier for each row

### `type`

In [19]:
netflix.type

0         Movie
1       TV Show
2       TV Show
3       TV Show
4       TV Show
         ...   
8802      Movie
8803    TV Show
8804      Movie
8805      Movie
8806      Movie
Name: type, Length: 8807, dtype: object

In [20]:
netflix.type.unique()

array(['Movie', 'TV Show'], dtype=object)

In [25]:
(
    netflix
    .loc[lambda df_: df_.listed_in.str.contains('TV', case=False)]
    .type
    .unique()
)

array(['TV Show'], dtype=object)

In [26]:
(
    netflix
    .loc[lambda df_: ~df_.listed_in.str.contains('TV', case=False)]
    .type
    .unique()
)

array(['Movie', 'TV Show'], dtype=object)

In [31]:
(
    netflix
    .loc[lambda df_: ~df_.listed_in.str.contains('TV', case=False)]
    .loc[lambda df_: df_.type.isin(['TV Show'])]
    .title
    .to_list()
)

['Turning Point: 9/11 and the War on Terror',
 'Top Secret UFO Projects: Declassified',
 'Transformers: War for Cybertron: Kingdom',
 'Revelations: The Masters of the Universe: Revelation Aftershow',
 'The Movies That Made Us',
 'Naomi Osaka',
 'How to Become a Tyrant',
 'Cat People',
 'Dogs',
 'The Last Dance',
 'High on the Hog: How African American Cuisine Transformed America',
 'Money, Explained',
 'Garth Brooks: The Road I’m On',
 'My Love: Six Stories of True Love',
 "DOTA: Dragon's Blood",
 'Formula 1: Drive to Survive',
 'Last Chance U: Basketball',
 'Pacific Rim: The Black',
 'Amend: The Fight for America',
 'We Are: The Brooklyn Saints',
 'Pretend It’s a City',
 'QB1: Beyond the Lights',
 'Transformers: War for Cybertron: Earthrise',
 'Song Exploder',
 'The Holiday Movies That Made Us',
 'Leah Remini: Scientology and the Aftermath',
 'Blood of Zeus',
 'My Next Guest Needs No Introduction With David Letterman',
 'Deaf U',
 'Whose Vote Counts, Explained',
 'The Chef Show',
 'Th

- It has two cardinality i.e. `Movie` and `TV Show`
- No null values
- `type` column has been verified, all data entries are correct.

### `title`

In [33]:
netflix.title

0        Dick Johnson Is Dead
1               Blood & Water
2                   Ganglands
3       Jailbirds New Orleans
4                Kota Factory
                ...          
8802                   Zodiac
8803              Zombie Dumb
8804               Zombieland
8805                     Zoom
8806                   Zubaan
Name: title, Length: 8807, dtype: object

In [38]:
(
    netflix
    .title
    .loc[lambda ser: ser.str.contains('[^A-Za-z0-9 ]+')]
)

1                                           Blood & Water
6                        My Little Pony: A New Generation
10                    Vendetta: Truth, Lies and The Mafia
14                        Crime Stories: India Detectives
16      Europe's Most Dangerous Man: Otto Skorzeny in ...
                              ...                        
8787                            You Can’t Fight Christmas
8790                        You Don't Mess with the Zohan
8793                                 Yours, Mine and Ours
8794                                               اشتباك
8795                                      Yu-Gi-Oh! Arc-V
Name: title, Length: 2533, dtype: object

In [107]:
netflix.loc[
    lambda df_: df_.title
                .str.strip()
                .str.title()
                .duplicated(keep=False)
].sort_values(['title', 'type', 'release_year'])

show_id     type                     title         director  \
3371   s3372    Movie              Consequences     Ozan Açıktan   
6529   s6530    Movie             Consequences      Ozan Açıktan   
5751   s5752  TV Show                DEATH NOTE              NaN   
5318   s5319    Movie                Death Note     Adam Wingard   
6705   s6706    Movie      Esperando La Carroza  Alejandro Doria   
303     s304    Movie      Esperando la carroza  Alejandro Doria   
5033   s5034    Movie       FullMetal Alchemist    Fumihiko Sori   
5095   s5096  TV Show       Fullmetal Alchemist              NaN   
7345   s7346    Movie            Love In A Puff   Pang Ho-cheung   
159     s160    Movie            Love in a Puff   Pang Ho-cheung   
8022   s8023  TV Show  Sin Senos sí Hay Paraíso              NaN   
1270   s1271  TV Show  Sin senos sí hay paraíso              NaN   

                                                   cast  \
3371  Nehir Erdoğan, Tardu Flordun, İlker Kaleli, Se...   
6529  Nehir Erdoğan, Tardu Flordun, İlker Kaleli, Se...   
5751  Kappei Yamaguchi, Alessandro Juliani, Mamoru M...   
5318  Willem Dafoe, Nat Wolff, Lakeith Stanfield, Ma...   
6705  Luis Brandoni, China Zorrilla, Antonio Gasalla...   
303   Luis Brandoni, China Zorrilla, Antonio Gasalla...   
5033  Ryosuke Yamada, Tsubasa Honda, Dean Fujioka, M...   
5095  Romi Park, Rie Kugimiya, Megumi Toyoguchi, Miy...   
7345  Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...   
159   Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...   
8022  Majida Issa, Fabián Ríos, Catherine Siachoque,...   
1270  Catherine Siachoque, Fabián Ríos, Carolina Gai...   

                      country         date_added  release_year rating  \
3371                   Turkey   October 25, 2019          2014  TV-MA   
6529                   Turkey   October 25, 2019          2014  TV-MA   
5751                    Japan   October 15, 2016          2006  TV-14   
5318            United States    August 25, 2017          2017  TV-MA   
6705                Argentina      July 15, 2018          1985     NR   
303                 Argentina     August 5, 2021          1985  TV-MA   
5033                    Japan  February 19, 2018          2017  TV-PG   
5095                    Japan    January 1, 2018          2003  TV-14   
7345                Hong Kong     August 1, 2018          2010  TV-MA   
159                 Hong Kong  September 1, 2021          2010  TV-MA   
8022  United States, Colombia   January 11, 2019          2018  TV-MA   
1270  United States, Colombia  February 25, 2021          2018  TV-MA   

       duration                                          listed_in  \
3371    106 min            Dramas, International Movies, Thrillers   
6529    106 min            Dramas, International Movies, Thrillers   
5751   1 Season  Anime Series, Crime TV Shows, International TV...   
5318    100 min                           Horror Movies, Thrillers   
6705     95 min        Comedies, Cult Movies, International Movies   
303      95 min        Comedies, Cult Movies, International Movies   
5033    134 min   Action & Adventure, Dramas, International Movies   
5095   1 Season               Anime Series, International TV Shows   
7345    103 min             Comedies, Dramas, International Movies   
159     103 min             Comedies, Dramas, International Movies   
8022  3 Seasons  International TV Shows, Spanish-Language TV Sh...   
1270  3 Seasons  International TV Shows, Spanish-Language TV Sh...   

                                            description  
3371  Secrets bubble to the surface after a sensual ...  
6529  Secrets bubble to the surface after a sensual ...  
5751  When a Japanese high schooler comes into posse...  
5318  Light Turner finds a supernatural notebook and...  
6705  Cora has three sons and a daughter and she´s a...  
303   Cora has three sons and a daughter and she´s a...  
5033  While alchemist Edward Elric searches for a wa...  
5095  Brothers Edward and A

In [97]:
(
    netflix
    .title
    .drop(index=[6529, 303, 159, 8022])
    .replace('Death Note', 'Death Note (2017)')
    .replace('FullMetal Alchemist', 'FullMetal Alchemist (2017)')
    .str.strip()
    .str.title()
    .loc[lambda ser: ser.duplicated(keep=False)]
)

Series([], Name: title, dtype: object)

In [98]:
(
    netflix
    .title
    .replace('Death Note', 'Death Note (2017)')
    .replace('FullMetal Alchemist', 'FullMetal Alchemist (2017)')
    .str.strip()
    .str.title()
)

0        Dick Johnson Is Dead
1               Blood & Water
2                   Ganglands
3       Jailbirds New Orleans
4                Kota Factory
                ...          
8802                   Zodiac
8803              Zombie Dumb
8804               Zombieland
8805                     Zoom
8806                   Zubaan
Name: title, Length: 8807, dtype: object

- indexes to drop: [6529, 303, 159, 8022]

### `director`

In [93]:
netflix.director

0       Kirsten Johnson
1                   NaN
2       Julien Leclercq
3                   NaN
4                   NaN
             ...       
8802      David Fincher
8803                NaN
8804    Ruben Fleischer
8805       Peter Hewitt
8806        Mozez Singh
Name: director, Length: 8807, dtype: object

In [94]:
netflix.director.unique()

array(['Kirsten Johnson', nan, 'Julien Leclercq', ..., 'Majid Al Ansari',
       'Peter Hewitt', 'Mozez Singh'], shape=(4529,), dtype=object)

In [117]:
netflix.director.isna().sum()

np.int64(2634)

In [116]:
(
    netflix
    .director
    .str.strip()
    .str.title()
    .fillna('unknown')
)

0       Kirsten Johnson
1               unknown
2       Julien Leclercq
3               unknown
4               unknown
             ...       
8802      David Fincher
8803            unknown
8804    Ruben Fleischer
8805       Peter Hewitt
8806        Mozez Singh
Name: director, Length: 8807, dtype: object

- Nothing can be done here

### `cast`

In [100]:
netflix.cast

0                                                     NaN
1       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...
2       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...
3                                                     NaN
4       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...
                              ...                        
8802    Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...
8803                                                  NaN
8804    Jesse Eisenberg, Woody Harrelson, Emma Stone, ...
8805    Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...
8806    Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...
Name: cast, Length: 8807, dtype: object

In [134]:
(
    netflix
    .cast
    .apply(lambda x: [] if pd.isna(x) else x.split(', '))
)

0                                                      []
1       [Ama Qamata, Khosi Ngema, Gail Mabalane, Thaba...
2       [Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nab...
3                                                      []
4       [Mayur More, Jitendra Kumar, Ranjan Raj, Alam ...
                              ...                        
8802    [Mark Ruffalo, Jake Gyllenhaal, Robert Downey ...
8803                                                   []
8804    [Jesse Eisenberg, Woody Harrelson, Emma Stone,...
8805    [Tim Allen, Courteney Cox, Chevy Chase, Kate M...
8806    [Vicky Kaushal, Sarah-Jane Dias, Raaghav Chana...
Name: cast, Length: 8807, dtype: object

### `country`

In [135]:
netflix.country

0       United States
1        South Africa
2                 NaN
3                 NaN
4               India
            ...      
8802    United States
8803              NaN
8804    United States
8805    United States
8806            India
Name: country, Length: 8807, dtype: object

In [137]:
netflix.country.unique()

array(['United States', 'South Africa', nan, 'India',
       'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia',
       'United Kingdom', 'Germany, Czech Republic', 'Mexico', 'Turkey',
       'Australia', 'United States, India, France', 'Finland',
       'China, Canada, United States',
       'South Africa, United States, Japan', 'Nigeria', 'Japan',
       'Spain, United States', 'France', 'Belgium',
       'United Kingdom, United States', 'United States, United Kingdom',
       'France, United States', 'South Korea', 'Spain',
       'United States, Singapore', 'United Kingdom, Australia, France',
       'United Kingdom, Australia, France, United States',
       'United States, Canada', 'Germany, United States',
       'South Africa, United States', 'United States, Mexico',
       'United States, Italy, France, Japan',
       'United States, Italy, Romania, United Kingdom',
       'Australia, United States', 'Argentina, Venezuela',
       'United States, United Kin

In [208]:
(
    netflix
    .country
    .str.findall('[A-Za-z]+ [A-Za-z]+ [A-Za-z]+|[A-Za-z]+ [A-Za-z]+|[A-Za-z]+')
    .explode()
    .unique()
)

array(['United States', 'South Africa', nan, 'India', 'Ghana',
       'Burkina Faso', 'United Kingdom', 'Germany', 'Ethiopia',
       'Czech Republic', 'Mexico', 'Turkey', 'Australia', 'France',
       'Finland', 'China', 'Canada', 'Japan', 'Nigeria', 'Spain',
       'Belgium', 'South Korea', 'Singapore', 'Italy', 'Romania',
       'Argentina', 'Venezuela', 'Hong Kong', 'Russia', 'Ireland',
       'Nepal', 'New Zealand', 'Brazil', 'Greece', 'Jordan', 'Colombia',
       'Switzerland', 'Israel', 'Taiwan', 'Bulgaria', 'Algeria', 'Poland',
       'Saudi Arabia', 'Thailand', 'Indonesia', 'Egypt', 'Denmark',
       'Kuwait', 'Netherlands', 'Malaysia', 'Vietnam', 'Hungary',
       'Sweden', 'Lebanon', 'Syria', 'Philippines', 'Iceland',
       'United Arab Emirates', 'Norway', 'Qatar', 'Mauritius', 'Austria',
       'Cameroon', 'Palestine', 'Uruguay', 'Kenya', 'Chile', 'Luxembourg',
       'Cambodia', 'Bangladesh', 'Portugal', 'Cayman Islands', 'Senegal',
       'Serbia', 'Malta', 'Namibia', '

In [223]:
(
    netflix
    .country
    .str.findall('[A-Za-z]+ [A-Za-z]+ [A-Za-z]+|[A-Za-z]+ [A-Za-z]+|[A-Za-z]+')
)

0       [United States]
1        [South Africa]
2                   NaN
3                   NaN
4               [India]
             ...       
8802    [United States]
8803                NaN
8804    [United States]
8805    [United States]
8806            [India]
Name: country, Length: 8807, dtype: object

### `date_added`

In [217]:
netflix.date_added

0       September 25, 2021
1       September 24, 2021
2       September 24, 2021
3       September 24, 2021
4       September 24, 2021
               ...        
8802     November 20, 2019
8803          July 1, 2019
8804      November 1, 2019
8805      January 11, 2020
8806         March 2, 2019
Name: date_added, Length: 8807, dtype: object

In [220]:
pd.to_datetime(netflix.date_added, format='mixed')

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8807, dtype: datetime64[ns]

### `release_year`

In [222]:
netflix.release_year

0       2020
1       2021
2       2021
3       2021
4       2021
        ... 
8802    2007
8803    2018
8804    2009
8805    2006
8806    2015
Name: release_year, Length: 8807, dtype: int64

### `rating`

In [225]:
netflix.rating

0       PG-13
1       TV-MA
2       TV-MA
3       TV-MA
4       TV-MA
        ...  
8802        R
8803    TV-Y7
8804        R
8805       PG
8806    TV-14
Name: rating, Length: 8807, dtype: object

In [226]:
netflix.rating.unique()

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', '74 min', '84 min', '66 min', 'NR', nan,
       'TV-Y7-FV', 'UR'], dtype=object)

In [228]:
(
    netflix
    .loc[lambda df_: df_.rating.isin(['74 min', '84 min', '66 min'])]
)

show_id   type                                 title    director  \
5541   s5542  Movie                       Louis C.K. 2017  Louis C.K.   
5794   s5795  Movie                 Louis C.K.: Hilarious  Louis C.K.   
5813   s5814  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.   

            cast        country          date_added  release_year  rating  \
5541  Louis C.K.  United States       April 4, 2017          2017  74 min   
5794  Louis C.K.  United States  September 16, 2016          2010  84 min   
5813  Louis C.K.  United States     August 15, 2016          2015  66 min   

     duration listed_in                                        description  
5541      NaN    Movies  Louis C.K. muses on religion, eternal love, gi...  
5794      NaN    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
5813      NaN    Movies  The comic puts his trademark hilarious/thought...

In [260]:
# minute_rows = netflix['rating'].str.contains('min', na=False)

# # Perform the shift
# netflix.loc[minute_rows, 'duration'] = netflix.loc[minute_rows, 'rating']
# netflix.loc[minute_rows, 'rating'] = np.nan

In [234]:
netflix.rating.unique()

array(['PG-13', 'TV-MA', 'PG', 'TV-14', 'TV-PG', 'TV-Y', 'TV-Y7', 'R',
       'TV-G', 'G', 'NC-17', nan, 'NR', 'TV-Y7-FV', 'UR'], dtype=object)

### `duration`

In [230]:
netflix.duration

0          90 min
1       2 Seasons
2        1 Season
3        1 Season
4       2 Seasons
          ...    
8802      158 min
8803    2 Seasons
8804       88 min
8805       88 min
8806      111 min
Name: duration, Length: 8807, dtype: object

In [235]:
(
    netflix
    .duration
)

0          90 min
1       2 Seasons
2        1 Season
3        1 Season
4       2 Seasons
          ...    
8802      158 min
8803    2 Seasons
8804       88 min
8805       88 min
8806      111 min
Name: duration, Length: 8807, dtype: object

In [236]:
(
    netflix
    .duration
    .unique()
)

array(['90 min', '2 Seasons', '1 Season', '91 min', '125 min',
       '9 Seasons', '104 min', '127 min', '4 Seasons', '67 min', '94 min',
       '5 Seasons', '161 min', '61 min', '166 min', '147 min', '103 min',
       '97 min', '106 min', '111 min', '3 Seasons', '110 min', '105 min',
       '96 min', '124 min', '116 min', '98 min', '23 min', '115 min',
       '122 min', '99 min', '88 min', '100 min', '6 Seasons', '102 min',
       '93 min', '95 min', '85 min', '83 min', '113 min', '13 min',
       '182 min', '48 min', '145 min', '87 min', '92 min', '80 min',
       '117 min', '128 min', '119 min', '143 min', '114 min', '118 min',
       '108 min', '63 min', '121 min', '142 min', '154 min', '120 min',
       '82 min', '109 min', '101 min', '86 min', '229 min', '76 min',
       '89 min', '156 min', '112 min', '107 min', '129 min', '135 min',
       '136 min', '165 min', '150 min', '133 min', '70 min', '84 min',
       '140 min', '78 min', '7 Seasons', '64 min', '59 min', '139 min',
    

In [248]:
(
    netflix
    .loc[lambda df_: 
    df_.duration.str.extract(r'([0-9]+) min', expand=False)
    .astype(float)
    .le(20)
    ]
)

show_id   type                                              title  \
71       s72  Movie                        A StoryBots Space Adventure   
694     s695  Movie                                              Aziza   
695     s696  Movie                                     Besieged Bread   
991     s992  Movie                                        One Like It   
1425   s1426  Movie                        Creating The Queen's Gambit   
1484   s1485  Movie                                   Cops and Robbers   
1557   s1558  Movie                                             Canvas   
1669   s1670  Movie                     If Anything Happens I Love You   
1745   s1746  Movie  Ya no estoy aquí: Una conversación entre Guill...   
1962   s1963  Movie                            A Love Song for Latasha   
2109   s2110  Movie                  John Was Trying to Contact Aliens   
2248   s2249  Movie                             The Claudia Kishi Club   
2713   s2714  Movie                                        Sol Levante   
2830   s2831  Movie                            Sitara: Let Girls Dream   
2858   s2859  Movie  Calico Critters: Everyone's Big Dream Flying i...   
3007   s3008  Movie                                  WHAT DID JACK DO?   
3364   s3365  Movie  The Road to El Camino: Behind the Scenes of El...   
3366   s3367  Movie                                   Little Miss Sumo   
3535   s3536  Movie   American Factory: A Conversation with the Obamas   
3709   s3710  Movie                                              ANIMA   
3775   s3776  Movie                     Cosmos Laundromat: First Cycle   
3777   s3778  Movie                                             Silent   
4707   s4708  Movie                                               Zion   
4892   s4893  Movie                                      Pocoyo & Cars   
5375   s5376  Movie                Buddy Thunderstruck: The Maybe Pile   
5456   s5457  Movie                             Michael Lost and Found   
6405   s6406  Movie                  Calico Critters: A Town of Dreams   
8205   s8206  Movie                               The Battle of Midway   

                               director  \
71                      David A. Vargas   
694                      Soudade Kaadan   
695                      Soudade Kaadan   
991                        Marwan Nabil   
1425                                NaN   
1484     Arnon Manor, Timothy Ware-Hill   
1557                 Frank E. Abney III   
1669     Will McCormack, Michael Govier   
1745                                NaN   
1962               Sophia Nahli Allison   
2109                                NaN   
2248                           Sue Ding   
2713                       Akira Saitoh   
2830              Sharmeen Obaid-Chinoy   
2858                      Momoko Kamiya   
3007                        David Lynch   
3364                                NaN   
3366                           Matt Kay   
3535                                NaN   
3709               Paul Thomas Anderson   
3775                     Mathieu Auvray   
3777  Limbert Fabian, Brandon Oldenburg   
4707                         Floyd Russ   
4892                                NaN   
5375                      Harry Chaskin   
5456                      Daniel Wilner   
6405                      Momoko Kamiya   
8205                          John Ford   

                                                   cast  \
71    Evan Spiridellis, Erin Fitzgerald, Jeff Gill, ...   
694                  Caress Bashar, Abdel Moneim Amayri   
695            Lama Hakeim, Gabriel Malki, Ehab Shaaban   
991   Riham Abdel Ghafour, Mohammed Aly Rizk, Hazem ...   
1425                                                NaN   
1484                                  Timothy Ware-Hill   
1557                                                NaN   
1669                                                NaN   
1745                 Guillermo del Toro, Alfonso Cuarón   
1962                

In [250]:
(
    netflix
    .assign(
        movie_duration=lambda df_: (
            df_
            .duration
            .str.extract(r'([0-9]+) min', expand=False)
            .astype(float)
        ),
        tv_show_seasons=lambda df_: (
            df_
            .duration
            .str.extract(r'([0-9]+) Seasons?', expand=False)
            .astype(float)
        )
    )
)

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  movie_duration  \
0     As her father nears the end of his life, filmm...            90.0   
1     After crossing paths at a party, a Cape Town t...             NaN   
2     To protect his family from a powerful drug lor...             NaN   
3     Feuds, flirtations and toilet talk go down amo...             NaN   
4     In a city of coaching centers known to train I...             NaN   
...                                                 ...             ...   
8802  A political cartoonist, a crime reporter and a...           158.0   
8803  While living alone in a spooky town, a young g...             NaN   
8804  Looking to survive in a world taken over by zo...            88.0   
8805  Dragged from civilian life, a former superhero...            88.0   
8806  A scrappy but poor boy worms his way into a ty...           111.0   

    

### `listed_in`

In [251]:
netflix.listed_in

0                                           Documentaries
1         International TV Shows, TV Dramas, TV Mysteries
2       Crime TV Shows, International TV Shows, TV Act...
3                                  Docuseries, Reality TV
4       International TV Shows, Romantic TV Shows, TV ...
                              ...                        
8802                       Cult Movies, Dramas, Thrillers
8803               Kids' TV, Korean TV Shows, TV Comedies
8804                              Comedies, Horror Movies
8805                   Children & Family Movies, Comedies
8806       Dramas, International Movies, Music & Musicals
Name: listed_in, Length: 8807, dtype: object

In [252]:
(
    netflix
    .listed_in
    .str.split(', ')
    .explode()
    .unique()
)

array(['Documentaries', 'International TV Shows', 'TV Dramas',
       'TV Mysteries', 'Crime TV Shows', 'TV Action & Adventure',
       'Docuseries', 'Reality TV', 'Romantic TV Shows', 'TV Comedies',
       'TV Horror', 'Children & Family Movies', 'Dramas',
       'Independent Movies', 'International Movies', 'British TV Shows',
       'Comedies', 'Spanish-Language TV Shows', 'Thrillers',
       'Romantic Movies', 'Music & Musicals', 'Horror Movies',
       'Sci-Fi & Fantasy', 'TV Thrillers', "Kids' TV",
       'Action & Adventure', 'TV Sci-Fi & Fantasy', 'Classic Movies',
       'Anime Features', 'Sports Movies', 'Anime Series',
       'Korean TV Shows', 'Science & Nature TV', 'Teen TV Shows',
       'Cult Movies', 'TV Shows', 'Faith & Spirituality', 'LGBTQ Movies',
       'Stand-Up Comedy', 'Movies', 'Stand-Up Comedy & Talk Shows',
       'Classic & Cult TV'], dtype=object)

In [253]:
(
    netflix
    .listed_in
    .str.split(', ')
)

0                                         [Documentaries]
1       [International TV Shows, TV Dramas, TV Mysteries]
2       [Crime TV Shows, International TV Shows, TV Ac...
3                                [Docuseries, Reality TV]
4       [International TV Shows, Romantic TV Shows, TV...
                              ...                        
8802                     [Cult Movies, Dramas, Thrillers]
8803             [Kids' TV, Korean TV Shows, TV Comedies]
8804                            [Comedies, Horror Movies]
8805                 [Children & Family Movies, Comedies]
8806     [Dramas, International Movies, Music & Musicals]
Name: listed_in, Length: 8807, dtype: object

### `description`

In [254]:
netflix.description

0       As her father nears the end of his life, filmm...
1       After crossing paths at a party, a Cape Town t...
2       To protect his family from a powerful drug lor...
3       Feuds, flirtations and toilet talk go down amo...
4       In a city of coaching centers known to train I...
                              ...                        
8802    A political cartoonist, a crime reporter and a...
8803    While living alone in a spooky town, a young g...
8804    Looking to survive in a world taken over by zo...
8805    Dragged from civilian life, a former superhero...
8806    A scrappy but poor boy worms his way into a ty...
Name: description, Length: 8807, dtype: object

## 5. Cleaning Operations

In [255]:
index_to_drop = [6529, 303, 159, 8022]

In [ ]:
minute_rows = netflix['rating'].str.contains('min', na=False)

# Perform the shift
netflix.loc[minute_rows, 'duration'] = netflix.loc[minute_rows, 'rating']
netflix.loc[minute_rows, 'rating'] = np.nan

In [264]:
def clean_data(df):
    return (
        df
        .drop(index=index_to_drop)
        .drop_duplicates()
        .assign(**{
            col: df[col].str.strip()
            for col in df.select_dtypes(include='O').columns
        })
        .rename(columns=str.lower)
        .assign(
            title=lambda df_: (
                df_
                .title
                .replace('Death Note', 'Death Note (2017)')
                .replace('FullMetal Alchemist', 'FullMetal Alchemist (2017)')
                .str.strip()
                .str.title()
            ),
            director=lambda df_: (
                df_
                .director
                .str.strip()
                .str.title()
                .fillna('unknown')
            ),
            cast=lambda df_: (
                df_
                .cast
                .apply(lambda x: [] if pd.isna(x) else x.split(', '))
            ),
            country=lambda df_: (
                df_
                .country
                .str.findall('[A-Za-z]+ [A-Za-z]+ [A-Za-z]+|[A-Za-z]+ [A-Za-z]+|[A-Za-z]+')
            ),
            date_added=lambda df_: pd.to_datetime(df_.date_added, format='mixed'),
            movie_duration=lambda df_: (
                df_
                .duration
                .str.extract(r'([0-9]+) min', expand=False)
                .astype(float)
            ),
            tv_show_seasons=lambda df_: (
                df_
                .duration
                .str.extract(r'([0-9]+) Seasons?', expand=False)
                .astype(float)
            ),
            listed_in=lambda df_: (
                df_
                .listed_in
                .str.split(', ')
            )
        )
        .drop(columns=['duration'])
    )
clean_data(netflix)

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water          unknown   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans          unknown   
4         s5  TV Show           Kota Factory          unknown   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb          unknown   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast          country  \
0                                                    []  [United States]   
1     [Ama Qamata, Khosi Ngema, Gail Mabalane, Thaba...   [South Africa]   
2     [Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nab...              NaN   
3                                                    []              NaN   
4     [Mayur More, Jitendra Kumar, Ranjan Raj, Alam ...          [India]   
...                                                 ...              ...   
8802  [Mark Ruffalo, Jake Gyllenhaal, Robert Downey ...  [United States]   
8803                                                 []              NaN   
8804  [Jesse Eisenberg, Woody Harrelson, Emma Stone,...  [United States]   
8805  [Tim Allen, Courteney Cox, Chevy Chase, Kate M...  [United States]   
8806  [Vicky Kaushal, Sarah-Jane Dias, Raaghav Chana...          [India]   

     date_added  release_year rating  \
0    2021-09-25          2020  PG-13   
1    2021-09-24          2021  TV-MA   
2    2021-09-24          2021  TV-MA   
3    2021-09-24          2021  TV-MA   
4    2021-09-24          2021  TV-MA   
...         ...           ...    ...   
8802 2019-11-20          2007      R   
8803 2019-07-01          2018  TV-Y7   
8804 2019-11-01          2009      R   
8805 2020-01-11          2006     PG   
8806 2019-03-02          2015  TV-14   

                                              listed_in  \
0                                       [Documentaries]   
1     [International TV Shows, TV Dramas, TV Mysteries]   
2     [Crime TV Shows, International TV Shows, TV Ac...   
3                              [Docuseries, Reality TV]   
4     [International TV Shows, Romantic TV Shows, TV...   
...                                                 ...   
8802                   [Cult Movies, Dramas, Thrillers]   
8803           [Kids' TV, Korean TV Shows, TV Comedies]   
8804                          [Comedies, Horror Movies]   
8805               [Children & Family Movies, Comedies]   
8806   [Dramas, International Movies, Music & Musicals]   

                                            description  movie_duration  \
0     As her father nears the end of his life, filmm...            90.0   
1     After crossing paths at a party, a Cape Town t...             NaN   
2     To protect his family from a powerful drug lor...             NaN   
3     Feuds, flirtations and toilet talk go down amo...             NaN   
4     In a city of coaching centers known to train I...             NaN   
...                                                 ...             ...   
8802  A political cartoonist, a crime reporter and a...           158.0   
8803  While living alone in a spooky town, a young g...             NaN   
8804  Looking to survive in a world taken over by zo...            88.0   
8805  Dragged from civilian life, a former superhero...            88.0   
8806  A scrappy but poor boy worms his way into a ty...           111.0   

      tv_show_seasons  
0                 NaN  
1                 2.0  
2                 1.0  
3                 1.0  
4                 2.0  
...               ...  
8802              NaN  
8803              2.0  
8804

In [265]:
netflix_cleaned = clean_data(netflix)

In [266]:
netflix_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8803 entries, 0 to 8806
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   show_id          8803 non-null   object        
 1   type             8803 non-null   object        
 2   title            8803 non-null   object        
 3   director         8803 non-null   object        
 4   cast             8803 non-null   object        
 5   country          7972 non-null   object        
 6   date_added       8793 non-null   datetime64[ns]
 7   release_year     8803 non-null   int64         
 8   rating           8796 non-null   object        
 9   listed_in        8803 non-null   object        
 10  description      8803 non-null   object        
 11  movie_duration   6128 non-null   float64       
 12  tv_show_seasons  2675 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(9)
memory usage: 962.8+ KB


## 7. Export the data

In [268]:
def export_data(data, name):
    file_name = f'{name}.csv'
    file_path = os.path.join(PROJECT_DIR, DATA_DIR, file_name)
    data.to_csv(file_path, index=False)
    return pd.read_csv(file_path).head()

In [269]:
export_data(netflix_cleaned, 'netflix_cleaned')

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water          unknown   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans          unknown   
4      s5  TV Show           Kota Factory          unknown   

                                                cast            country  \
0                                                 []  ['United States']   
1  ['Ama Qamata', 'Khosi Ngema', 'Gail Mabalane',...   ['South Africa']   
2  ['Sami Bouajila', 'Tracy Gotoas', 'Samuel Jouy...                NaN   
3                                                 []                NaN   
4  ['Mayur More', 'Jitendra Kumar', 'Ranjan Raj',...          ['India']   

   date_added  release_year rating  \
0  2021-09-25          2020  PG-13   
1  2021-09-24          2021  TV-MA   
2  2021-09-24          2021  TV-MA   
3  2021-09-24          2021  TV-MA   
4  2021-09-24          2021  TV-MA   

                                           listed_in  \
0                                  ['Documentaries']   
1  ['International TV Shows', 'TV Dramas', 'TV My...   
2  ['Crime TV Shows', 'International TV Shows', '...   
3                       ['Docuseries', 'Reality TV']   
4  ['International TV Shows', 'Romantic TV Shows'...   

                                         description  movie_duration  \
0  As her father nears the end of his life, filmm...            90.0   
1  After crossing paths at a party, a Cape Town t...             NaN   
2  To protect his family from a powerful drug lor...             NaN   
3  Feuds, flirtations and toilet talk go down amo...             NaN   
4  In a city of coaching centers known to train I...             NaN   

   tv_show_seasons  
0              NaN  
1              2.0  
2              1.0  
3              1.0  
4              2.0